python/ql/test/library-tests/examples/custom-sanitizer/Taint.qll

```diff
@@ -0,0 +1,73 @@
+import python
+import semmle.python.security.TaintTracking
+import semmle.python.security.strings.Untrusted
+
+class SimpleSource extends TaintSource {
+    SimpleSource() { this.(NameNode).getId() = "TAINTED_STRING" }
+
+    override predicate isSourceOf(TaintKind kind) { kind instanceof ExternalStringKind }
+
+    override string toString() { result = "taint source" }
+}
+
+class MySimpleSanitizer extends Sanitizer {
+    MySimpleSanitizer() { this = "MySimpleSanitizer" }
+
+    /**
+     * The test `if is_safe(arg):` sanitizes `arg` on its `true` edge.
+     *
+     * Can't handle `if not is_safe(arg):` :\ that's why it's called MySimpleSanitizer
+     *
+     */
+    override predicate sanitizingEdge(TaintKind taint, PyEdgeRefinement test) {
+        taint instanceof ExternalStringKind and
+        exists(CallNode call |
+            test.getTest() = call and test.getSense() = true
+        |
+            call = Value::named("test.is_safe").getACall() and
+            test.getInput().getAUse() = call.getAnArg()
+        )
+    }
+}
+
+class MySanitizerHandlingNot extends Sanitizer {
+    MySanitizerHandlingNot() { this = "MySanitizerHandlingNot" }
+
+    /** The test `if is_safe(arg):` sanitizes `arg` on its `true` edge. */
+    override predicate sanitizingEdge(TaintKind taint, PyEdgeRefinement test) {
+        taint instanceof ExternalStringKind and
+        exists(CallNode call |
+            clears_taint_on_true(call, test.getTest(), test.getSense())
+        |
+            call = Value::named("test.is_safe").getACall() and
+            test.getInput().getAUse() = call.getAnArg()
+        )
+    }
+
+    private predicate clears_taint_on_true(ControlFlowNode final_test, ControlFlowNode test, boolean sense) {
+        final_test = test and
+        sense = true
```

Comment:
I'm not entirely happy about this predicate. It will (absent whatever magic gets applied) hold at least for all tuples `(n, n, true)` where `n` is any `ControlFlowNode`, and this seems like a pretty big overapproximation.

Perhaps you could do the same thing as in the [TarSlip sanitizer](https://github.com/Semmle/ql/pull/2888/files#diff-2083315c64b2aefb4bb9e365fb17364dR130) and push the context that uses `call`/`final_test` inside this predicate, thus eliminating the `final_test` argument?

Reply:
yep, that is a very good point! I understanding how to think about QL evaluation better and better :+1: